arangod/GeneralServer/H2CommTask.cpp

```
@@ -0,0 +1,654 @@
+////////////////////////////////////////////////////////////////////////////////
+
+#include "H2CommTask.h"
+
+#include "Basics/Exceptions.h"
+#include "Basics/ScopeGuard.h"
+#include "Basics/asio_ns.h"
+#include "Cluster/ServerState.h"
+#include "GeneralServer/GeneralServer.h"
+#include "GeneralServer/GeneralServerFeature.h"
+#include "Logger/LogMacros.h"
+#include "Rest/HttpRequest.h"
+#include "Rest/HttpResponse.h"
+#include "Statistics/ConnectionStatistics.h"
+#include "Statistics/RequestStatistics.h"
+
+#include <velocypack/velocypack-aliases.h>
+#include <cstring>
+
+using namespace arangodb;
+using namespace arangodb::basics;
+using namespace arangodb::rest;
+
+using arangodb::velocypack::StringRef;
+
+template <SocketType T>
+/*static*/ int H2CommTask<T>::on_begin_headers(nghttp2_session* session,
+                                               const nghttp2_frame* frame, void* user_data) {
+  H2CommTask<T>* me = static_cast<H2CommTask<T>*>(user_data);
+
+  if (frame->hd.type != NGHTTP2_HEADERS || frame->headers.cat != NGHTTP2_HCAT_REQUEST) {
+    return 0;
+  }
+
+  int32_t sid = frame->hd.stream_id;
+  auto req = std::make_unique<HttpRequest>(me->_connectionInfo, /*messageId*/ sid,
+                                           /*allowMethodOverride*/ false);
+  me->createStream(sid, std::move(req));
+
+  return 0;
+}
+
+template <SocketType T>
+/*static*/ int H2CommTask<T>::on_header(nghttp2_session* session, const nghttp2_frame* frame,
+                                        const uint8_t* name, size_t namelen,
+                                        const uint8_t* value, size_t valuelen,
+                                        uint8_t flags, void* user_data) {
+  H2CommTask<T>* me = static_cast<H2CommTask<T>*>(user_data);
+  int32_t stream_id = frame->hd.stream_id;
+
+  if (frame->hd.type != NGHTTP2_HEADERS || frame->headers.cat != NGHTTP2_HCAT_REQUEST) {
+    return 0;
+  }
+
+  Stream* strm = me->findStream(stream_id);
+  if (!strm) {
+    return 0;
```

should we return an error code from here?

---

I don't know I am following the example code